### ISSUE TITLE
Using foundation for apps in an existing angular project

### ISSUE THREAD

**Issue by laurent-le-graverend** [zurb/foundation-apps#428]
In an existing AngularJS project, I've installed foundation for apps using:

```bower install foundation-apps --save```

In foundation's bower.json file there are only these 2 files to be included:

```
"main": [
  "scss/foundation.scss",
  "js/angular/app.js"
],
```

I am using `<!-- bower:js -->` to include automatically all the files from bower dependencies. So only js/angular/app.js will be loaded (and maybe not necessary since I already bootstrap my app and define the routing?).

This obviously misses to load all necessary JS files to fully use Foundation.
What is the clean way to load the JavasScript files from Foundation?

**Comment by joshuajohnson814** [comment]
+1

**Comment by gakimball** [comment]
We can just add those missing files to the `bower.json`. Should that property also include every Sass file?

**Comment by joshuajohnson814** [comment]
I think for the javascript there should be a distribution file that concats all of the javascript files similar to how Angular-UI Bootstrap packages their bower release. In regards to the CSS I think the same approach should be taken which is to build a foundation.css. Something like this:

"main": [
  "dist/foundation-apps.css",
  "dist/foundation-apps.js"
]

This way anyone who adds this bower dependency can get it working out of the box with just those two files. If anyone wishes to override the foundation SCSS files with _settings they can choose to ignore the foundation-apps.css and create their own.

**Comment by joshuajohnson814** [comment]
I don't think you should expose the actual app.js file through bower. If someone wishes to use this framework as a bower dependency then they will already have their own Angular app file. Thoughts?

**Comment by circlingthesun** [comment]
@joshuajohnson814 +1

**Comment by gakimball** [comment]
@joshuajohnson814 I agree, and in fact it shouldn't be there anymore. Our documentation uses a different entry file, and the template stack also uses it's own. If you're incorporating the source files into your own build process, that file is unnecessary, although it does serve as a useful template.

As for including distribution files, we can look at it for the next patch. We're considering including a `dist` folder on the same repo, for this and perhaps also Foundation for Sites, to avoid the hassle of having a separate repo just for the Bower install.

**Comment by joshuajohnson814** [comment]
@gakimball I definitely think having the app.js file in there adds value as a template like you said. 

Currently I'm using this bower module in my project and I'm referencing the SCSS files directly so that I can override the styling in my own process. It will be awesome when I can just reference the foundation-apps.js file directly and not have to build those out in my process.

I'm really excited about the future of this framework! Great work guys.

**Comment by gakimball** [comment]
@joshuajohnson814 The main issue we've had with the file is that some folks assumed it was a module to be included as a dependency, when it's just meant to be boilerplate. Also, since it's in the `bower_components` folder, which means it probably isn't in version control, you can't edit it.

*But* we haven't heard anything recently, so we can let it sit for now :)

**Comment by gakimball** [comment]
So for JavaScript you just add every file, but how does it work with Sass files? I was looking at wiredep (which is what @laurent-le-graverend is using) but I couldn't figure out exactly how it works with Sass. Since Sass files are pre-compiled, it's not like you're concatenating them all together; you just need an import path. So I'm wondering if the `main` property of `bower.json` needs every Sass file or just the entry point.

**Comment by joshuajohnson814** [comment]
You shouldn't be adding every file to main in bower. You should add the dist versions in the form of a single js file and CSS file.

**Comment by gakimball** [comment]
I see, so it's not for source files. Thanks! We'll set up a `dist` folder for the next patch.

**Comment by joshuajohnson814** [comment]
Yeah exactly. Wiredep is looking for the files it can take and inject straight into index.HTML. I like that you have the source in bower though so we can override

**Comment by laurent-le-graverend** [comment]
Having a `dist` is definitely a must when we install it with bower as a dependency, so everyone can quickly start with all necessary files.

Then if they want to customise their build and have only specific modules (JS and SCSS) from Foundation for Apps, they can use Gulp or Grunt to exclude `dist` files and include what they need.

This last part would be nice to be documented later, using Gulp for example to make clear for everyone that they have different options to work with Foundation.

**Comment by gakimball** [comment]
Alright, this commit adds a Gulp task to create a `dist` folder with minified and un-minified CSS and JS: e6aab4e097abafce1ccce6586f0bd5d7726dde97

And this commit is me forgetting to update the `main` entry in `bower.json` ;) f5e33b0d3abf8d3817555c7b5a7361902794f1a7

We're trying to put out 1.0.3 this week, which will be the first version with the `dist` folder.

**Comment by joshuajohnson814** [comment]
I noticed that you were including the bower_components dependencies in the dist JS file. The dist should only include foundation specific JS files.

https://github.com/zurb/foundation-apps/pull/433

**Comment by gakimball** [comment]
@joshuajohnson814 Ah, okay. In that case, what would be the best way to package the other dependencies for distribution? Maybe just bundle all of the external libraries into one file?

Obviously wiredep helps with this dependency resolution, but not everyone will be using that. Some folks will just want to drop in finished files to get going.

**Comment by joshuajohnson814** [comment]
You shouldn't package the dependencies for distribution. Your bower.json file lists the dependencies so when a user adds foundation-apps with bower it will automatically install the dependencies for them.

Take a look at my pull request. I separated the bower components into a separate array just to load for testing. The gulp watch task should only be watching your foundation-app files and gulp dist should only distribute foundation-app files.

**Comment by circlingthesun** [comment]
I can't speak for others but I have no need for 3rd party dependencies to
be bundled with foundation. In general its often problematic if they are.
Surely people that have trouble with dependencies are the same people that
will be using the foundation boilerplate which do deal with them?

**Comment by joshuajohnson814** [comment]
I thought of something else.... In order to properly distribute foundation-apps.js you will also need to add the component HTML files as angular templates to the javascript file using templateCache.

This is exactly how Angular UI Bootstrap distributes their framework.

https://github.com/angular-ui/bootstrap-bower

**Comment by laurent-le-graverend** [comment]
+1 for the templates into the JS directly, they are really small HTML files and it's better to avoid unnecessary http requests.

But the distribution with bower also needs to have uncompiled SCSS and JS files in order to let you make your own 'build' (select only the JS and SCSS components you needs).

**Comment by joshuajohnson814** [comment]
Bower Recap:
dist/foundation-app.js, dist/foundation-app.min.js (concat all foundation angular JS files except app.js with HTML templates using templateCache)
dist/foundation-app.css, dist/foundation-app.min.css
js/
scss/ 

Do you want me to open up a separate issue for this request?

**Comment by laurent-le-graverend** [comment]
How about this (similar as jQuery)

* bower.json (`"main"` will reference both JS and CSS min files)
* dist/
  * foundation-app.css
  * foundation-app.min.css
  * foundation-app.js
  * foundation-app.min.js
* src/
  * js/ 
  * scss/

Then, in your own top module

### Choice 1: use all components (using min files) ###

```
  angular
    .module('myApp', [
      'foundation',
      'ui.router',
      ...
  ])
```

### Choice 2: pick your own ###

```
  angular
    .module('myApp', [
      'foundation.core',
      'foundation.common',
      'foundation.offcanvas',
      ...,
      'ui.router',
      ...
  ])
```

Last thing to solve for this second choice would be also to use have components templates in the JS templateCache.
This could be a part of your own build documented.

**Comment by gakimball** [comment]
@laurent-le-graverend The way you described module loading already works with our current file structure. Originally you had to add each UI component's module separately, but in version 1.0.2 we added a `foundation` module which is just a shorthand for all of the UI components.

But as far as the compilation step, it sounds like the dist file should just be *our* framework files, and not any of the five or so libraries we rely on. So we'll make that change. (This also brings into stark relief how many dependencies this library has right now. Maybe for version 2 we'll look at separating modules that have dependencies, like our touch stuff.)

As far as using template caching, we've gotten that request before, so we'll look into it.

**Comment by gakimball** [comment]
Merged in #433 which addresses this. Thanks for your help, friends!
